source/common/router/weighted_cluster_specifier.cc

```diff
@@ -0,0 +1,321 @@
+      // compute the selected value. So, we treat multi-valued header as invalid input and fall back
+      // to use internally generated random number.
+      uint64_t random_value = 0;
+      if (absl::SimpleAtoi(header_value[0]->value().getStringView(), &random_value)) {
+        random_value_from_header = random_value;
+      }
+    }
+
+    if (!random_value_from_header.has_value()) {
+      // Random value should be found here. But if it is not set due to some errors, log the
+      // information and fallback to the random value that is set by stream id.
+      ENVOY_LOG(debug, "The random value can not be found from the header and it will fall back to "
+                       "the value that is set by stream id");
+    }
+  }
+
+  const bool runtime_key_prefix_configured = !runtime_key_prefix_.empty();
+  uint32_t total_cluster_weight = total_cluster_weight_;
+  absl::InlinedVector<uint32_t, 4> cluster_weights;
+
+  // if runtime config is used, we need to recompute total_weight
+  if (runtime_key_prefix_configured) {
+    // Temporary storage to hold consistent cluster weights. Since cluster weight
+    // can be changed with runtime keys, we need a way to gather all the weight
+    // and aggregate the total without a change in between.
+    // The InlinedVector will be able to handle at least 4 cluster weights
+    // without allocation. For cases when more clusters are needed, it is
+    // reserved to ensure at most a single allocation.
+    cluster_weights.reserve(weighted_clusters_.size());
+
+    total_cluster_weight = 0;
+    for (const auto& cluster : weighted_clusters_) {
+      auto cluster_weight = cluster->clusterWeight(loader_);
+      cluster_weights.push_back(cluster_weight);
+      if (cluster_weight > std::numeric_limits<uint32_t>::max() - total_cluster_weight) {
+        IS_ENVOY_BUG("Sum of weight cannot overflow 2^32");
+        return nullptr;
+      }
+      total_cluster_weight += cluster_weight;
+    }
+  }
+
+  if (total_cluster_weight == 0) {
+    IS_ENVOY_BUG("Sum of weight cannot be zero");
+    return nullptr;
+  }
+  const uint64_t selected_value =
+      (random_value_from_header.has_value() ? random_value_from_header.value() : random_value) %
+      total_cluster_weight;
+  uint64_t begin = 0;
+  uint64_t end = 0;
+  auto cluster_weight = cluster_weights.begin();
+
+  // Find the right cluster to route to based on the interval in which
+  // the selected value falls. The intervals are determined as
+  // [0, cluster1_weight), [cluster1_weight, cluster1_weight+cluster2_weight),..
+  for (const auto& cluster : weighted_clusters_) {
+
+    if (runtime_key_prefix_configured) {
+      end = begin + *cluster_weight++;
+    } else {
+      end = begin + cluster->clusterWeight(loader_);
+    }
+
+    if (selected_value >= begin && selected_value < end) {
+      if (!cluster->cluster_name_.empty()) {
+        return std::make_shared<WeightedClusterEntry>(std::move(parent), "", cluster);
```

Comment:
`WeightedClusterEntry` seems to be creating a new route entry for each request. Just wondering if this could be expensive under high load.